flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java

```diff
@@ -0,0 +1,573 @@
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.core.Is.is;
+import static org.hamcrest.core.IsNot.not;
+import static org.junit.Assert.fail;
+
+/** {@code PackagedProgramRetrieverImplTest} tests {@link PackagedProgramRetrieverImpl}. */
+public class PackagedProgramRetrieverImplTest extends TestLogger {
+
+    @Rule public ClasspathProvider classpathProvider = new ClasspathProvider();
+
+    @Test
+    public void testDeriveEntryClassInformationForPythonBasedOnJobName() throws FlinkException {
+        // TODO: see FLINK-23154
+        Assume.assumeTrue(
+                "This tests only succeeds if " + ConfigConstants.ENV_FLINK_OPT_DIR + " is set.",
+                System.getProperty(ConfigConstants.ENV_FLINK_OPT_DIR) != null);
+        final EntryClassInformationProvider informationProvider =
+                PackagedProgramRetrieverImpl.createEntryClassInformationProvider(
+                        null,
+                        Collections.emptyList(),
+                        null,
+                        PackagedProgramUtils.getPythonDriverClassName(),
+                        new String[0]);
+        assertThat(informationProvider.getJobClassName().isPresent(), is(true));
+        assertThat(
+                informationProvider.getJobClassName().get(),
+                is(PackagedProgramUtils.getPythonDriverClassName()));
+        assertThat(informationProvider.getJarFile().isPresent(), is(true));
+        assertThat(
+                informationProvider.getJarFile().get(),
+                is(PackagedProgramUtils.getPythonJar().getFile()));
+    }
+
+    @Test
+    public void testDeriveEntryClassInformationForPythonBasedOnParameter() throws FlinkException {
+        // TODO: see FLINK-23154
+        Assume.assumeTrue(
+                "This tests only succeeds if " + ConfigConstants.ENV_FLINK_OPT_DIR + " is set.",
+                System.getProperty(ConfigConstants.ENV_FLINK_OPT_DIR) != null);
+
+        final EntryClassInformationProvider informationProvider =
+                PackagedProgramRetrieverImpl.createEntryClassInformationProvider(
+                        null, Collections.emptyList(), null, null, new String[] {"--python"});
+
+        assertThat(informationProvider.getJobClassName().isPresent(), is(true));
+        assertThat(
+                informationProvider.getJobClassName().get(),
+                is(PackagedProgramUtils.getPythonDriverClassName()));
+        assertThat(informationProvider.getJarFile().isPresent(), is(true));
+        assertThat(
+                informationProvider.getJarFile().get(),
+                is(PackagedProgramUtils.getPythonJar().getFile()));
+    }
+
+    @Test
+    public void testDeriveEntryClassInformationForCustomJar()
+            throws FlinkException, MalformedURLException {
+        // make loading from system classpath fail to make sure that it's not triggered
+        classpathProvider.setSystemClasspathWithTwoEntryClasses();
```

Comment:
I don't understand why this would make it fail, nor what "it" that is not triggered actually is. Wouldn't it be more obvious if you did something like clearing the classpath?

Reply:
Good idea, that's more explicit. I agree.